[et] generate debug handle before opeartor decomposition

Gasoonjia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #11997

This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

Gasoonjia committed:
[et] generate debug handle before opeartor decomposition

This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[et] generate debug handle before opeartor decomposition"


This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[et] generate debug handle before opeartor decomposition"


This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[et] generate debug handle before opeartor decomposition"


This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

[ghstack-poisoned]

kimishpatel commented on exir/passes/debug_handle_generator_pass.py:
```diff
@@ -4,31 +4,75 @@
+            """
+
+            node_source = node.meta[FROM_NODE_KEY]
+            node_source = node_source[-1]
```
this is only populated for unlifted graph when `.module` is called? if so how is it guranteed that this is present

Gasoonjia replied:
no it is not only populate that `.module` is explicitly called. In our case it is guarantee generated after run_decompostition so we will have that when generating debug handle.

kimishpatel replied:
> it is guarantee generated after run_decompostition so we will have that when generating debug handle.

There in my question around to_edge_transform_and_lower. although that specifically is unrelated to this pr

kimishpatel reviewed (commented):
For the most part looks good. I left a couple of comments, one specifically around relation of to_edge_lower_and_transform to debug handle (not related ot this diff)

Gasoonjia commented on exir/passes/debug_handle_generator_pass.py:
```diff
@@ -4,31 +4,75 @@
+
+            greatest_ancestor_node_id = _get_greatest_ancestor_node_identifier(node)
+
+            debug_handle = (
+                len(source_node_id_to_debug_handle) + 1
+                if greatest_ancestor_node_id not in source_node_id_to_debug_handle
+                else source_node_id_to_debug_handle[greatest_ancestor_node_id]
+            )
```
For first question, after from the currnet numeric descrepency detector we are working on, there's no other tools or infra that rely on our debug handle. There're some testcases rely on debug handle for check and I've talked with the owners and they are ok with the update. 

For second question, not sure if I fully understand your concern, but our current debug handle generation pass also locates at the end of to_edge. Our update does not change the location. But we do have concern about whether or not the change of debug handle generation may impact the correctness of the debug handels in the downstream e.g. in delegate, though ci is all clean. And we will pay attention to it when working on the numeric descrepancy dtector on delegated model.

kimishpatel replied:
Actually we should discuss this second paragraph further. THe issue is that `to_edge_tranform_and_lower` happens before we run decomposition, is that right? If so you dont have debug handles generated no? If no debug handles then no debug information. I would like to be either wrong about my assumption or we need to fix it

kimishpatel replied:
For the first paragraph, do you forsee any disruption in existing user flows for debugging? That would be one thing that would make me block this PR, so having an answer on that would be nice

Gasoonjia committed:
Update on "[et] generate debug handle before opeartor decomposition"


This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

[ghstack-poisoned]

Gasoonjia commented on devtools/inspector/_inspector_utils.py:
```diff
@@ -303,14 +303,18 @@ def gen_graphs_from_etrecord(
     return op_graph_map
 
 
+# One debug handle should only be associated with one node. We are in the middle of migrating debug handle generation
+# from graph after to_edge to graph after torch.export, one every debug handle in exported graph may be associated with multiple nodes in to_edge
```
i've updated the comment to reflect the function behavior before, during and after the migration.

Gasoonjia committed:
Update on "[et] generate debug handle before opeartor decomposition"


This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[et] generate debug handle before opeartor decomposition"


This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[et] generate debug handle before opeartor decomposition"


This diff update the debug handle generation, from each node in the edge program having a individual debug handle, to all nodes having a same ancestor in export graph sharing a same debug handle, which update the start point of tracing our node transformation from edge graph to exported graph.

Differential Revision: [D76860368](https://our.internmc.facebook.com/intern/diff/D76860368/)

[ghstack-poisoned]
